java/tsfile/src/main/java/org/apache/tsfile/read/TsFileSequenceReader.java

```diff
@@ -1808,41 +1807,11 @@ public ByteBuffer readPage(PageHeader header, CompressionType type) throws IOExc
     if (header.getUncompressedSize() == 0) {
       return buffer;
     }
-    if (decryptor == null || decryptor.getEncryptionType() == EncryptionType.UNENCRYPTED) {
-      if (type == CompressionType.UNCOMPRESSED) {
-        return buffer;
-      } else {
-        IUnCompressor unCompressor = IUnCompressor.getUnCompressor(type);
-        ByteBuffer uncompressedBuffer = ByteBuffer.allocate(header.getUncompressedSize());
-        unCompressor.uncompress(
-            buffer.array(), buffer.position(), buffer.remaining(), uncompressedBuffer.array(), 0);
-        return uncompressedBuffer;
-      }
-    } else {
-      if (type == CompressionType.UNCOMPRESSED) {
-        ByteBuffer decryptedBuffer = ByteBuffer.allocate(header.getUncompressedSize());
-        System.arraycopy(
-            decryptor.decrypt(buffer.array(), buffer.position(), buffer.remaining()),
-            0,
-            decryptedBuffer.array(),
-            0,
-            buffer.remaining());
-        return decryptedBuffer;
-      } else {
-        ByteBuffer decryptedBuffer = ByteBuffer.allocate(header.getCompressedSize());
-        System.arraycopy(
-            decryptor.decrypt(buffer.array(), buffer.position(), buffer.remaining()),
-            0,
-            decryptedBuffer.array(),
-            0,
-            buffer.remaining());
-        IUnCompressor unCompressor = IUnCompressor.getUnCompressor(type);
-        ByteBuffer uncompressedBuffer = ByteBuffer.allocate(header.getUncompressedSize());
-        unCompressor.uncompress(
-            decryptedBuffer.array(), 0, buffer.remaining(), uncompressedBuffer.array(), 0);
-        return uncompressedBuffer;
-      }
-    }
+    byte[] decryptedData = decryptor.decrypt(buffer.array(), buffer.position(), buffer.remaining());
+    IUnCompressor unCompressor = IUnCompressor.getUnCompressor(type);
+    ByteBuffer uncompressedBuffer = ByteBuffer.allocate(header.getUncompressedSize());
+    unCompressor.uncompress(decryptedData, 0, decryptedData.length, uncompressedBuffer.array(), 0);
+    return uncompressedBuffer;
   }
```

Comment:
This does not avoid copying, I guess.
```java
   ByteBuffer finalBuffer = decrypt(decryptor, buffer);
   finalBuffer = uncompress(compressionType, buffer, header.getUncompressedSize());
    return finalBuffer;
  }
  
  private static ByteBuffer decrypt(IDecryptor decryptor, ByteBuffer buffer) {
      if (decryptor == null || decryptor.getEncryptionType() == EncryptionType.UNENCRYPTED) {
          return buffer;
      }
      return ByteBuffer.wrap(decryptor.decrypt(buffer.array(), buffer.arrayOffset() + buffer.position(), buffer.remaining()));
  }
  
  private static ByteBuffer uncompress(CompressionType compressionType, ByteBuffer buffer, int uncompressedSize) {
      if (compressionType == UNCOMPRESSED) {
          return buffer;
      }
      IUnCompressor unCompressor = IUnCompressor.getUnCompressor(type);
      ByteBuffer uncompressedBuffer = ByteBuffer.allocate(uncompressedSize);
        unCompressor.uncompress(
            buffer.array(), buffer.arrayOffset() + buffer.position(), buffer.remaining(), uncompressedBuffer.array(), 0);
        return uncompressedBuffer;
  }
```

The main idea is that a sequential structure is much easier to understand than a nesting structure. Try making your code simpler.

Reply:
You may put the static methods in another place and share them with other functions that need them, if any.
Beware of the offset when directly using the underlying array from a ByteBuffer. It should be ByteBuffer.arrayOffset() + ByteBuffer.position() instead of ByteBuffer.position().